### PR TITLE
[PyTorch]Add aten::linear and fuse addmm to linear

### PR DESCRIPTION
**Pull request by zrphercule**
Summary:

Documentation:

[Optional Fixes #issue]

Test Plan:

Please see a detailed explanation of how to fill out the fields in the relevant sections in PULL_REQUEST.md.

### CODE REVIEW

File: torch_glow/src/PyTorchModelLoader.cpp [hunk 1]
```diff
@@ -314,6 +314,35 @@ void PyTorchModelLoader::loadConvolution(const torch::jit::Node *ptNode) {
...
+	  weights = 1,
+	  bias = 2,
+	}
+  }
+
+  glow::NodeValue input = getGlowNodeValue(inputs[Inputs::input]);
+  glow::NodeValue weights = getGlowNodeValue(inputs[Inputs::weights]);
+  glow::NodeValue bias = getGlowNodeValue(inputs[Inputs::bias]);
+
+  glow::NodeValue output;
```
Comment (jackm321) [inline torch_glow/src/PyTorchModelLoader.cpp]: Since there's no branching here, you can just use the result of `createAdd` below as the input to `addGlowNodeValue`

File: torch_glow/src/PyTorchModelLoader.cpp [hunk 2]
```diff
@@ -314,6 +314,35 @@ void PyTorchModelLoader::loadConvolution(const torch::jit::Node *ptNode) {
   addGlowNodeValue(outputs[0], output->getResult());
 }
 
+void PyTorchModelLoader::loadLinear(const torch::jit::Node *ptNode) {
+  auto inputs = ptNode->inputs();
+  auto outputs = ptNode->outputs();
+  assert(inputs.size() == 3);
+  assert(outputs.size() == 1);
+
+  // Indexes of aten::linear inputs
```
Comment (jackm321) [inline torch_glow/src/PyTorchModelLoader.cpp]: period

File: torch_glow/src/FusingOptimizer.h [hunk 3]
```diff
@@ -0,0 +1,7 @@
+#pragma once
+
+#include <troch/csrc/jit/ir.h>
```
Comment (jackm321) [inline torch_glow/src/FusingOptimizer.h]: `torch`

File: torch_glow/src/FusingOptimizer.cpp [hunk 4]
```diff
@@ -0,0 +1,49 @@
+#include "FusingOptimizer.h"
```
Comment (jackm321) [inline torch_glow/src/FusingOptimizer.cpp]: * Add Copyright
* note that this was orignially from `pytorch/tvm` (right?)
* Is this something that should/could be added to PyTorch directly instead?

File: torch_glow/src/FusingOptimizer.h [hunk 5]
```diff
@@ -0,0 +1,7 @@
...
+
+using namespace torch::jit;
```
Comment (jackm321) [inline torch_glow/src/FusingOptimizer.h]: don't do this in header

File: torch_glow/src/FusingOptimizer.h [hunk 6]
```diff
@@ -0,0 +1,7 @@
+#pragma once
```
Comment (jackm321) [inline torch_glow/src/FusingOptimizer.h]: We don't use #pragma once in glow

File: torch_glow/src/PyTorchModelLoader.cpp [hunk 7]
```diff
@@ -314,6 +314,35 @@ void PyTorchModelLoader::loadConvolution(const torch::jit::Node *ptNode) {
...
+	enum {
+	  input = 0,
+	  weights = 1,
+	  bias = 2,
+	}
+  }
+
+  glow::NodeValue input = getGlowNodeValue(inputs[Inputs::input]);
+  glow::NodeValue weights = getGlowNodeValue(inputs[Inputs::weights]);
+  glow::NodeValue bias = getGlowNodeValue(inputs[Inputs::bias]);
```
Comment (jackm321) [inline torch_glow/src/PyTorchModelLoader.cpp]: Bias is optional right? We should check `hasGlowNodeValue` first

File: torch_glow/src/PyTorchModelLoader.cpp [hunk 8]
```diff
@@ -314,6 +314,35 @@ void PyTorchModelLoader::loadConvolution(const torch::jit::Node *ptNode) {
   addGlowNodeValue(outputs[0], output->getResult());
 }
 
+void PyTorchModelLoader::loadLinear(const torch::jit::Node *ptNode) {
```
Comment (jackm321) [inline torch_glow/src/PyTorchModelLoader.cpp]: Linear can broadcast right? If so we should handle this.

File: torch_glow/src/PyTorchModelLoader.cpp [hunk 9]
```diff
@@ -314,6 +314,35 @@ void PyTorchModelLoader::loadConvolution(const torch::jit::Node *ptNode) {
...
+
+  glow::NodeValue input = getGlowNodeValue(inputs[Inputs::input]);
+  glow::NodeValue weights = getGlowNodeValue(inputs[Inputs::weights]);
+  glow::NodeValue bias = getGlowNodeValue(inputs[Inputs::bias]);
+
+  glow::NodeValue output;
+
+  // add some shape check here
+  // auto *broadcasted_bias = f_->createBroadcast(bias, {input.dims()});
+  auto *mul = f_->createMatmul("mul", input, weights);
```
Comment (jackm321) [inline torch_glow/src/PyTorchModelLoader.cpp]: make this something like "linear_mul" so we kinda know which op this was loaded from.

File: torch_glow/src/binding.cpp [hunk 10]
```diff
@@ -69,7 +70,11 @@ void registerGlowOp() {
 void registerPass() {
   torch::jit::RegisterPass pass([](std::shared_ptr<torch::jit::Graph> &g) {
     if (fusionPassEnabled) {
-      torch::jit::CustomFuseGraph(g, PyTorchModelLoader::isNodeSupported,
+
+	  // Fuse all linear operators
```
Comment (jackm321) [inline torch_glow/src/binding.cpp]: Add some more comments on what this means and why it's here.

File: torch_glow/src/PyTorchModelLoader.cpp (hunk 9 above)
Comment (jackm321) [inline torch_glow/src/PyTorchModelLoader.cpp]: This is basically an FC right? Can we just load this as an FC? Otherwise what is the point of fusing these together in the JIT IR?

**Comment by zrphercule** [comment]
Thanks for your review @jackm321 !

File: torch_glow/src/FusingOptimizer.cpp (hunk 4 above)
Comment (zrphercule) [reply torch_glow/src/FusingOptimizer.cpp]: Yeah this is copied from tvm. This is actually a temporary solution, since soon or later they will have linear support inside pytorch.

File: torch_glow/src/PyTorchModelLoader.cpp (hunk 7 above)
Comment (zrphercule) [reply torch_glow/src/PyTorchModelLoader.cpp]: yes it is in linear, although in addmm it is not, but you are right it is better to check since this is linear.

File: torch_glow/src/CMakeLists.txt [hunk 11]
```diff
@@ -14,7 +14,8 @@ link_directories(${PYTORCH_DIR}/lib)
 pybind11_add_module(_torch_glow
                       binding.cpp
                       PyTorchFileLoader.cpp
-                      PyTorchModelLoader.cpp
+					  FusingOptimizer.cpp
```
Comment (jackm321) [inline torch_glow/src/CMakeLists.txt]: weird indentation?

**Comment by zrphercule** [comment]
> Is this still WIP?

not anymore. I will rebase & clean it up, targeting merge it today lol.

File: torch_glow/src/CMakeLists.txt [hunk 12]
```diff
@@ -14,7 +14,8 @@ link_directories(${PYTORCH_DIR}/lib)
 pybind11_add_module(_torch_glow
                       binding.cpp
                       PyTorchFileLoader.cpp
-                      PyTorchModelLoader.cpp
+                      FusingOptimizer.cpp
+					  PyTorchModelLoader.cpp
```
Comment (jackm321) [inline torch_glow/src/CMakeLists.txt]: weird indentation still?

File: torch_glow/src/FusingOptimizer.cpp [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "FusingOptimizer.h"
+#include <torch/csrc/jit/passes/subgraph_rewrite.h>
+
+// This is mainly copied from pytorch/tvm
+// This pass fuse the addmm or matmul + add generated by JIT back to linear
+// to allow direct support with tvm integration with Relay IR
```
Comment (jackm321) [inline torch_glow/src/FusingOptimizer.cpp]: Glow IR

File: torch_glow/src/FusingOptimizer.cpp [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+  %output = aten::matmul(%input, %weight_t)
+  return (%output))IR";
+    std::string fused_linear_bias_none = R"IR(
+graph(%input, %weight):
+  %bias: Tensor? = prim::Constant()
+  %res = aten::linear(%input, %weight, %bias)
+  return (%res))IR";
+
+
+    // replace addmm pattern to linear
```
Comment (jackm321) [inline torch_glow/src/FusingOptimizer.cpp]: why is this indented more?

File: torch_glow/src/PyTorchModelLoader.cpp [hunk 15]
```diff
@@ -490,6 +491,39 @@ PyTorchModelLoader::loadConvolution(const torch::jit::Node *ptNode) {
...
+    };
+  };
+
+  glow::NodeValue input;
+  ASSIGN_VALUE_OR_RETURN_ERR(input, getGlowNodeValue(inputs[Inputs::input]));
+  glow::NodeValue weights;
+  ASSIGN_VALUE_OR_RETURN_ERR(weights, getGlowNodeValue(inputs[Inputs::weights]));
+
+  // add some shape check here
+  // auto *broadcasted_bias = f_->createBroadcast(bias, {input.dims()});
```
Comment (jackm321) [inline torch_glow/src/PyTorchModelLoader.cpp]: dead code

File: torch_glow/tests/nodes/addmm_test.py [hunk 16]
```diff
@@ -0,0 +1,18 @@
+import torch
+
+from tests.utils import jitVsGlow
+
+def test_addmm():
```
Comment (jackm321) [inline torch_glow/tests/nodes/addmm_test.py]: Shouldn't we just test linear here?

Comment (zrphercule) [reply torch_glow/tests/nodes/addmm_test.py]: here what we test is actaully addmm, if this is correct then linear is correct as well.

File: torch_glow/src/FusingOptimizer.cpp (hunk 14 above)
Comment (zrphercule) [reply torch_glow/src/FusingOptimizer.cpp]: It is because the IR code is not part of this indent...

File: torch_glow/tests/nodes/addmm_test.py (hunk 16 above)
Comment (jackm321) [reply torch_glow/tests/nodes/addmm_test.py]: Yeah but we don't support addmm directly, only the addmm instances that can be lowered to from linear right?

**Comment by zrphercule** [comment]
test of add_batchnorm is down? hmmmm

File: torch_glow/src/CMakeLists.txt [hunk 17]
```diff
@@ -14,8 +14,9 @@ link_directories(${PYTORCH_DIR}/lib)
 pybind11_add_module(_torch_glow
                       binding.cpp
                       PyTorchFileLoader.cpp
+                      FusingOptimizer.cpp
                       PyTorchModelLoader.cpp
-                      CachingGraphRunner.cpp)
+                     CachingGraphRunner.cpp)
```
Comment (jackm321) [inline torch_glow/src/CMakeLists.txt]: still weird :(

Comment (zrphercule) [reply torch_glow/src/CMakeLists.txt]: lolololol

**Comment by jackm321** [comment]
> test of add_batchnorm is down? hmmmm

fixed now

**Comment by zrphercule** [comment]
@jackm321 mind give me a big approve?

File: torch_glow/tests/nodes/addmm_test.py (hunk 16 above)
Comment (zrphercule) [reply torch_glow/tests/nodes/addmm_test.py]: yeah... so you think we can have a seperate for nn.linear?

**Comment by zrphercule** [comment]
Yeah I actually have an idea of this hahah... let me retake sigmoid first (smile on my face)

**Comment by zrphercule** [comment]
> Yeah I actually have an idea of this hahah... let me retake sigmoid first (smile on my face)

also, dont worry I will fix the remain lint problems.
